app/filtering.js

```diff
@@ -220,6 +218,38 @@ function registerForBeforeRedirect (session, partition) {
   })
 }
 
+module.exports.considerRequestExceptions = (requestHeaders, url, firstPartyUrl, isPrivate) => {
+  const cookieSetting = module.exports.isResourceEnabled(appConfig.resourceNames.COOKIEBLOCK, firstPartyUrl, isPrivate)
+  if (cookieSetting) {
+    const parsedTargetUrl = urlParse(url || '')
+    const parsedFirstPartyUrl = urlParse(firstPartyUrl)
+
+    if (cookieSetting === 'blockAllCookies' ||
+      isThirdPartyHost(parsedFirstPartyUrl.hostname, parsedTargetUrl.hostname)) {
+      let hasCookieException = false
+      const firstPartyOrigin = getOrigin(firstPartyUrl)
+      const targetOrigin = getOrigin(url)
+      if (cookieExceptions.hasOwnProperty(firstPartyOrigin) &&
+          cookieExceptions[firstPartyOrigin] === targetOrigin &&
+          cookieSetting !== 'blockAllCookies') {
+        hasCookieException = true
+      }
+      // Clear cookie and referer on third-party requests
+      if (requestHeaders['Cookie'] &&
+          firstPartyOrigin !== pdfjsOrigin && !hasCookieException) {
+        requestHeaders['Cookie'] = undefined
+      }
+      if (cookieSetting !== 'blockAllCookies' &&
```

Comment:
i think the `cookieSetting !== 'blockAllCookies'` part is a bug in the original code. the referer should be blocked when the cookieSetting is 'blockAllCookies'.